text/0000-mir.md

```diff
@@ -0,0 +1,821 @@
+2. The CFG. The CFG tries to model the flow of execution as a graph
+   rather than a tree, to help analyses in dealing with complex
+   control-flow formed by things like loops, `break`, `continue`, etc.
+   This CFG is however inferior to the MIR in that it is only an
+   approximation of control-flow and does not include all the
+   information one would need to actually execute the program (for
+   example, for an `if` expression, the CFG would indicate that two
+   branches are possible, but would not contain enough information to
+   decide which branch to take).
+3. `ExprUseVisitor`. The `ExprUseVisitor` is designed to work in
+   conjunction with the CFG. It walks the AST and highlights actions
+   of interest to later analyses, such as borrows or moves. For each
+   such action, the analysis gets a callback indicating the point in
+   the CFG where the action occurred along with what
+   happened. Overloaded operators, method calls, and so forth are
+   "desugared" into their more primitive operations. This is
+   effectively a kind of MIR, but it is not complete enough to do
+   translation, since it focuses purely on borrows, moves, and other
+   things of interest to the safety checker.
+
+Each of these things were added in order to try and cope with the
+complexity of working directly on the AST. The CFG for example
+consolidates knowledge about control-flow into one piece of code,
+producing a data structure that can be easily interpreted. Similarly,
+the `ExprUseVisitor` consolidates knowledge of how to walk and
+interpret the current compiler representation.
+
+### Goals
+
+It is useful to think about what "knowledge" the MIR should
+encapsulate. Here is a listing of the kinds of things that should be
+explicit in the MIR and thus that downstream code won't have to
+re-encode in the form of repeated logic:
+
+- **Precise ordering of control-flow.** The CFG makes this very explicit,
+  and the individual statements and nodes in the MIR are very small
+  and detailed and hence nothing "interesting" happens in the middle
+  of an individual node with respect to control-flow.
+- **What needs to be dropped and when.** The set of data that needs to
+  be dropped and when is a fairly complex thing to calculate: you have
+  to know what's in scope, including temporary values and so forth.
+  In the MIR, all drops are explicit, including those that result from
+  panics and unwinding.
+- **How matches are desugared.** Reasoning about matches has been a
+  traditional source of complexity. Matches combine traversing types
+  with borrows, moves, and all sorts of other things, depending on the
+  precise patterns in use. This is all vastly simplified and explicit
+  in MIR.
+  
+One thing the current MIR does not make explicit as explicit as it
+could is when something is *moved*. For by-value uses of a value, the
+code must still consult the type of the value to decide if that is a
+move or not. This could be made more explicit in the IR.
```

Comment:
@eefriedman

> In a world with drop calls explicitly encoded into the MIR, whether something is moved as opposed to copied doesn't matter at all; either the value will be explicitly dropped, or it won't. This is true with either the current embedded drop flags or explicit stack-based drop flags. Or am I missing something?

For one thing, the MIR as I've described it thus far is allowed to DROP things that may have been moved. I'm assuming a later pass that determines precisely what needs to be dropped and inserts code to prevent double drops; this will be a type-based, control-flow-sensitive analysis, and hence it makes sense to do it after the MIR is built.